Fix skipping levels of parent folders for the content layout Combobox

unitmatrix opened the pull request:
This is in response to #16259. May also address #15250, and #14426 and possibly some other tickets related to content layout selection ComboBox on the Add New Torrent Dialog which is known to break the file paths with several levels of depth so that after a few clicks between the available options: "Original", "Create Subfolder", "Don't create subfolder" the root folders get stripped completely for all options.

The root folder finder has been modified to account for the full common path rather than only the top level folder.

Ex.:

Folder1/Folder2/Folder3/File1
Folder1/Folder2/Folder32/File2
Folder1/Folder2/Folder33/File1

will have a root folder of 'Folder1/Folder2', and not 'Folder1' as they have now.

This makes the "Don't create subfolder" option actually create no subfolders for those folders that have no root folder and will result in the following path list:

Folder3/File1
Folder32/File2
Folder33/File1

Fixes #16259

glassez commented:
>This is in response to https://github.com/qbittorrent/qBittorrent/issues/16259.

The "root folder" of the torrent has always designated the topmost common folder of all torrent files. And the author of #16259 reports precisely the fact that qBittorrent mistakenly skips several nested subfolders instead of only topmost one.

unitmatrix commented:
Thank you for clarifying the idea of the "root folder".
Removed changes for the "full root" finder so that "root folder" corresponds to the topmost common folder as expected.
Added a reset of the file paths on each content layout ComboBox change to fix the unexpected removal of the "root folder".
This fixes #16259

glassez commented on src/gui/addnewtorrentdialog.cpp:
```diff
@@ -546,6 +546,7 @@ void AddNewTorrentDialog::contentLayoutChanged(const int index)
     const auto contentLayout = ((index == 0)
                                 ? BitTorrent::detectContentLayout(m_torrentInfo.filePaths())
                                 : static_cast<BitTorrent::TorrentContentLayout>(index));
+    m_torrentParams.filePaths = m_torrentInfo.filePaths();
```
This will discard all file name changes made before, so it is unacceptable.

I'm analyzing the problem... but it has a low priority.

unitmatrix replied:
Good point. Modified it to avoid discarding file name changes made before. It should now ensure that the root folder had in fact been stripped before adding it back.

an0n666 commented:
This seems like an importantant fix. Does this require any testing or changes?

glassez commented:
Superseded by #16724.

turbodude commented:
This is an issue again in 4.4.5

PriitUring commented:
@turbodude

Read previous comments before commenting. It waa written before your comment that this pull request was supeseded by another pull request. Now if you took time to read, then this is outdated pull request and is not the issue report place and in the other pull request that this was supeseded by has written that it fixes the old issue and that it's changes with that different pull request has been added to v4.5.0 Milestones and changes got merged to master branch 7 months ago.

So try newest Master builds in GitHub or use official v4.5.0beta1 installers or create your own fork in GitHub to include various fixes, because unfortunately very few fixes, code changes or improvements added to current Master branch (v4.5.x) gets backported to v4.4.x.
